sw/airborne/firmwares/rotorcraft/guidance/guidance_indi.c

```diff
@@ -145,29 +164,49 @@ void guidance_indi_run(bool in_flight, float heading_sp) {
   float speed_sp_y = pos_y_err * guidance_indi_pos_gain;
   float speed_sp_z = pos_z_err * guidance_indi_pos_gain;
 
-  sp_accel.x = (speed_sp_x - stateGetSpeedNed_f()->x) * guidance_indi_speed_gain;
-  sp_accel.y = (speed_sp_y - stateGetSpeedNed_f()->y) * guidance_indi_speed_gain;
-  sp_accel.z = (speed_sp_z - stateGetSpeedNed_f()->z) * guidance_indi_speed_gain;
+  // If the acceleration setpoint is set over ABI message
+  if (indi_accel_sp_set_2d) {
+    sp_accel.x = indi_accel_sp.x;
+    sp_accel.y = indi_accel_sp.y;
+    // In 2D the vertical motion is derived from the flight plan
+    sp_accel.z = (speed_sp_z - stateGetSpeedNed_f()->z) * guidance_indi_speed_gain;
+    float dt = get_sys_time_float() - time_of_accel_sp_2d;
+    if (dt > 0.5) {
+      indi_accel_sp_set_2d = false;
```

Comment:
This seems strange... 

I am not sure where else 'sp_accel' used but is overwritten somewhere else? Else it should be sufficient to immediately set indi_accel_sp_set_2d = false; without the dt check.

Reply:
if the guidance control loop is faster than the setpoint update, you want to avoid to switch the 2D/3D control and the derivate from speed until a timeout is reached

Reply:
I remember that when we wrote this condition for dt was also for safety porpuses. In particular, if the commands from the ground stop, then the rotorcraft resumes the flight plan (for example, hovering in a position with the guidance from INDI). Indeed, if the setpoint update is slower than the guidance, you do not want to have a "misbehavior" by switching between two set points from different systems (formation control or guidance).

I agree with kirk that this part could be more clear... in fact, I do not even remember how we wrote it. Although the behavior was as expected (including the safety).

Reply:
Indeed, the setpoint is updated over telemetry, which will update much slower than the periodic frequency.

Reply:
merging with an extra comment